File: app/src/main/java/me/raatiniemi/worker/data/service/ResumeService.java
/*
 * Copyright (C) 2016 Worker Project
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package me.raatiniemi.worker.data.service;

import android.content.Intent;
import android.util.Log;

import java.util.Date;

import me.raatiniemi.worker.R;
import me.raatiniemi.worker.domain.interactor.ClockIn;
import me.raatiniemi.worker.domain.interactor.GetProject;
import me.raatiniemi.worker.domain.model.Project;
import me.raatiniemi.worker.presentation.notification.ErrorNotification;
import me.raatiniemi.worker.presentation.notification.PauseNotification;

public class ResumeService extends OngoingService {
    private static final String TAG = "ResumeService";

    public ResumeService() {
        super(TAG);
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        long projectId = getProjectId(intent);

        try {
            ClockIn clockIn = buildClockInUseCase();
            clockIn.execute(projectId, new Date());

            updateUserInterface(projectId);

            if (isOngoingNotificationEnabled()) {
                GetProject getProject = buildGetProjectUseCase();
                Project project = getProject.execute(projectId);

                sendPauseNotification(project);
                return;
            }

            dismissResumeNotification(projectId);
        } catch (Exception e) {
            Log.w(TAG, "Unable to resume project", e);

            sendErrorNotification(projectId);
        }
    }

    ClockIn buildClockInUseCase() {
        return new ClockIn(getTimeRepository());
    }

    GetProject buildGetProjectUseCase() {
        return new GetProject(getProjectRepository());
    }

    private void sendPauseNotification(Project project) {
        sendNotification(
                project.getId(),
                PauseNotification.build(this, project)
        );
    }

    private void dismissResumeNotification(long projectId) {
        dismissNotification(projectId);
    }

    private void sendErrorNotification(long projectId) {
        sendNotification(
                projectId,
                ErrorNotification.build(
                        this,
                        getString(R.string.error_notification_resume_title),
                        getString(R.string.error_notification_resume_message)
                )
        );
    }
}

File: app/src/main/java/me/raatiniemi/worker/feature/ongoing/service/ResumeService.kt
/*
 * Copyright (C) 2021 Tobias Raatiniemi
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package me.raatiniemi.worker.feature.ongoing.service

import android.content.Intent
import kotlinx.coroutines.*
import me.raatiniemi.worker.domain.project.model.Project
import me.raatiniemi.worker.domain.project.usecase.GetProject
import me.raatiniemi.worker.domain.time.Milliseconds
import me.raatiniemi.worker.domain.timeinterval.usecase.ActiveProjectException
import me.raatiniemi.worker.domain.timeinterval.usecase.CalculateTimeToday
import me.raatiniemi.worker.domain.timeinterval.usecase.ClockIn
import me.raatiniemi.worker.feature.ongoing.view.PauseNotification
import me.raatiniemi.worker.monitor.analytics.Event
import me.raatiniemi.worker.monitor.analytics.UsageAnalytics
import org.koin.android.ext.android.inject
import timber.log.Timber
import kotlin.coroutines.CoroutineContext

internal class ResumeService : OngoingService("ResumeService"), CoroutineScope {
    override val coroutineContext: CoroutineContext
        get() = Dispatchers.IO + Job()

    private val usageAnalytics: UsageAnalytics by inject()
    private val clockIn: ClockIn by inject()
    private val getProject: GetProject by inject()
    private val calculateTimeToday: CalculateTimeToday by inject()

    override fun onHandleIntent(intent: Intent?) {
        launch {
            try {
                val projectId = getProjectId(intent)
                val project = getProject(projectId)

                clockIn(project)

                val calculatedTimeToday = calculateTimeToday(project)
                withContext(Dispatchers.Main) {
                    updateUserInterface(project)
                    sendOrDismissPauseNotification(project, calculatedTimeToday)
                }
            } catch (e: Exception) {
                Timber.e(e, "Unable to resume project")
            }
        }
    }

    private suspend fun clockIn(project: Project) {
        try {
            clockIn(project, Milliseconds.now)

            usageAnalytics.log(Event.NotificationClockIn)
        } catch (e: ActiveProjectException) {
            Timber.w(e, "Resume service called with active project")
        }
    }

    private fun sendOrDismissPauseNotification(project: Project, calculatedTimeToday: Long) {
        sendOrDismissOngoingNotification(project) {
            PauseNotification.build(
                this,
                project,
                calculatedTimeToday,
                isOngoingNotificationChronometerEnabled
            )
        }
    }
}
